olp-cpp-sdk-core/src/client/PendingUrlRequests.cpp

```diff
@@ -0,0 +1,254 @@
+    cancelled_requests_.emplace(it->first, it->second);
+    pending_requests_.erase(it);
+  }
+
+  return ret;
+}
+
+bool PendingUrlRequests::CancelAll() {
+  // This only cancells the ongoing Network request the callback trigger
+  // is taking care of the Network callback.
+  std::lock_guard<std::mutex> lock(mutex_);
+  for (auto& request : pending_requests_) {
+    request.second->Cancel();
+  }
+
+  return true;
+}
+
+bool PendingUrlRequests::CancelAllAndWait() {
+  PendingRequestsType pending_requests;
+  PendingRequestsType cancelled_requests;
+
+  {
+    // Copy, do not move else the OnRequestCompleted callback
+    // will not work
+    std::lock_guard<std::mutex> lock(mutex_);
+    pending_requests = pending_requests_;
+    cancelled_requests = cancelled_requests_;
+  }
+
+  bool ret = true;
+
+  for (auto& request : pending_requests) {
+    if (request.second && !request.second->CancelAndWait()) {
+      OLP_SDK_LOG_WARNING_F(kLogTag,
+                            "CancelAllAndWait() timeout on pending url=%s",
+                            request.first.c_str());
+      ret = false;
+    }
+  }
+
+  for (auto& request : cancelled_requests) {
+    if (request.second && !request.second->CancelAndWait()) {
+      OLP_SDK_LOG_WARNING_F(kLogTag,
+                            "CancelAllAndWait() timeout on cancelled url=%s",
+                            request.first.c_str());
+      ret = false;
+    }
+  }
+
+  return ret;
+}
+
+PendingUrlRequests::PendingUrlRequestPtr PendingUrlRequests::operator[](
+    const std::string& url) {
+  PendingUrlRequestPtr result_ptr = nullptr;
+
+  // TODO: Check here that the pending request was not cancelled before
+  // adding new callback to it?
+
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    auto it = pending_requests_.find(url);
+    if (it != pending_requests_.end()) {
+      result_ptr = it->second;
+    } else {
+      result_ptr = std::make_shared<PendingUrlRequest>();
+      pending_requests_.emplace(url, result_ptr);
+    }
+  }
+
+  return result_ptr;
+}
+
+void PendingUrlRequests::OnRequestCompleted(http::RequestId request_id,
+                                            const std::string& url,
+                                            HttpResponse response) {
+  PendingUrlRequestPtr request_ptr;
+
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    auto pending_it = pending_requests_.find(url);
+    auto cancelled_it = cancelled_requests_.find(url);
+
+    // Either the request is in the pending or in the cancelled requests list.
+    // Check that it is the same request. It might happen that we have a
+    // cancelled request and then a followup valid request with the same URL.
+    if (pending_it != pending_requests_.end() &&
+        request_id == pending_it->second->GetRequestId()) {
+      request_ptr = std::move(pending_it->second);
+      pending_requests_.erase(pending_it);
+    } else if (cancelled_it != cancelled_requests_.end() &&
```

Comment:
May there be a situations where one url request has been cancelled and the second one with the same url is added?

Reply:
Yes. It might be that we one request cancels its callback but the Network takes some time to return the answer and there is another request from a different call for the same URL.